feat(core): add automatic migration from Renderer to Renderer2

crisbeto opened the pull request:
Adds a schematic and tslint rule that automatically migrate the consumer from `Renderer` to `Renderer2`. Supports:
* Renaming imports.
* Renaming property and method argument types.
* Casting to `Renderer`.
* Mapping all of the methods from the `Renderer` to `Renderer2`.

Note that some of the `Renderer` methods don't map cleanly between renderers. In these cases the migration adds a helper function at the bottom of the file which ensures that we generate valid code with the same return value as before. E.g. here's what the migration for `createText` looks like.

Before:
```
class SomeComponent {
  createAndAddText() {
    const node = this._renderer.createText(this._element.nativeElement, 'hello');
    node.textContent += ' world';
  }
}
```

After:
```
class SomeComponent {
  createAndAddText() {
    const node = __rendererCreateTextHelper(this._renderer, this._element.nativeElement, 'hello');
    node.textContent += ' world';
  }
}

function __rendererCreateTextHelper(renderer: any, parent: any, value: any) {
  const node = renderer.createText(value);
  if (parent) {
    renderer.appendChild(parent, node);
  }
  return node;
}
```

This PR resolves FW-1344.

crisbeto committed:
feat(core): add automatic migration from Renderer to Renderer2

Adds a schematic and tslint rule that automatically migrate the consumer from `Renderer` to `Renderer2`. Supports:
* Renaming imports.
* Renaming property and method argument types.
* Casting to `Renderer`.
* Mapping all of the methods from the `Renderer` to `Renderer2`.

Note that some of the `Renderer` methods don't map cleanly between renderers. In these cases the migration adds a helper function at the bottom of the file which ensures that we generate valid code with the same return value as before. E.g. here's what the migration for `createText` looks like.

Before:
```
class SomeComponent {
  createAndAddText() {
    const node = this._renderer.createText(this._element.nativeElement, 'hello');
    node.textContent += ' world';
  }
}
```

After:
```
class SomeComponent {
  createAndAddText() {
    const node = __rendererCreateTextHelper(this._renderer, this._element.nativeElement, 'hello');
    node.textContent += ' world';
  }
}

function __rendererCreateTextHelper(renderer: any, parent: any, value: any) {
  const node = renderer.createText(value);
  if (parent) {
    renderer.appendChild(parent, node);
  }
  return node;
}
```

This PR resolves FW-1344.

kara commented on packages/core/schematics/migrations/renderer-to-renderer2/google3/rendererToRenderer2Rule.ts:
```diff
@@ -0,0 +1,121 @@
+        ts.EmitHint.Unspecified, replaceImport(node, 'Renderer', 'Renderer2'), sourceFile);
+
+    return new RuleFailure(
+        sourceFile, node.getStart(), node.getEnd(), 'Imports of Renderer are not allowed.',
```
Can we add some info on what to do? like:

```suggestion
        sourceFile, node.getStart(), node.getEnd(), 'Imports of deprecated Renderer are not allowed. Please use Renderer2 instead. ',
```

kara commented on packages/core/schematics/migrations/renderer-to-renderer2/google3/rendererToRenderer2Rule.ts:
```diff
@@ -0,0 +1,121 @@
+    const type = node.type !;
+
+    return new RuleFailure(
+        sourceFile, type.getStart(), type.getEnd(), 'References to Renderer are not allowed.',
```
Same here

```suggestion
        sourceFile, type.getStart(), type.getEnd(), 'References to deprecated Renderer are not allowed. Please use Renderer2 instead.',
```

kara commented on packages/core/schematics/migrations/renderer-to-renderer2/helpers.ts:
```diff
@@ -0,0 +1,392 @@
+
+/** Names of the helper functions that are supported for this migration. */
+export const enum HelperFunction {
+  createElement = '__rendererCreateElementHelper',
```
The `__` prefix may still be too common. Seems like we should use something private to Angular like the theta or at least an "ng"

kara replied:
Actually, what do you think of `__ngRendererCreateElementHelper`? I like that the underscore shows that it is private (and then the prefix reduces collisions)

kara commented on packages/core/schematics/migrations/renderer-to-renderer2/helpers.ts:
```diff
@@ -0,0 +1,392 @@
+
+  // for (let i = 0; i < allNodes.length; i++) {
+  //   renderer.destroyNode(allNodes[i]);
+  // }
```
Thanks for all these comments - super helpful

devversion replied:
Maybe worth mentioning: we could technically also parse the TS source-code into the declaration code (which would be less verbose). e.g. 

```ts
const helper = ts.createSourceFile('', `
  for (let i = 0; i < allNodes.length; i++) {
    renderer.destroyNode(allNodes[i]);
  }
`, ts.ScriptTarget.Latest).statements[0];
```

Not sure what is best though. Constructing the nodes programmatically is also a valid option :smile:

kara commented on packages/core/schematics/migrations/renderer-to-renderer2/helpers.ts:
```diff
@@ -0,0 +1,392 @@
+  // throw new Error('...');
+  const throwStatement = ts.createThrow(ts.createNew(
+      ts.createIdentifier('Error'), [],
+      [ts.createStringLiteral('Renderer.animate is no longer supported!')]));
```
Hmm it feels a little weird that we are inserting a new method that will just throw. Maybe we should just comment out the `animate` call and add a `TODO: replace with Angular animations (Renderer.animate is no longer supported`

kara replied:
Ah, I didn't realize. Strange 🤔

kara commented on packages/core/schematics/test/renderer_to_renderer2_migration_spec.ts:
```diff
@@ -0,0 +1,1100 @@
+      await runMigration();
+
+      // Expect the `setInfo` method to only contain whitespace.
+      expect(tree.readContent('/index.ts')).toMatch(/setInfo\(\) \{\s+\}/);
```
Might be nice to have a comment here with: `TODO: remove setBindingDebugInfo call (as it is not supported)`

crisbeto replied:
We've already removed it for them though. I doubt that anybody's using it in the first place, because that call hasn't been doing anything for a long time now.

kara replied:
OK, just thinking it might be weird that code disappears without explanation. Maybe we can put it in the guide.

kara commented on packages/core/schematics/migrations/renderer-to-renderer2/index.ts:
```diff
@@ -0,0 +1,120 @@
+  return (tree: Tree) => {
+    const {buildPaths, testPaths} = getProjectTsConfigPaths(tree);
+    const basePath = process.cwd();
+    const allPaths = [...buildPaths, ...testPaths];
```
Seems like we should have some sort of header here logged by the migration, so users have context for any errors / failures that come next?  Similar to https://github.com/angular/angular/blob/master/packages/core/schematics/migrations/static-queries/index.ts#L57.

That way, we have an easy place to drop in a link to the migration guide later too

devversion replied:
Yes. Plus it allows us to investigate migration issues easier. People usually just create an issue which contains the error message since the CLI doesn't print out the migration stack trace. Therefore printing the header allows us to easily associate it with the given migration.

kara reviewed (changes requested):
Thanks for doing all this work! Seems like it is pretty close! Some thoughts below

kara commented on packages/core/schematics/test/renderer_to_renderer2_migration_spec.ts:
```diff
@@ -0,0 +1,1100 @@
+          }
+        }
+      `));
+    });
```
Do we have any tests that ensure that these helpers are printed in the right place? e.g. at the bottom of the file? Seems like we might want a sanity check.

crisbeto replied:
We didn't, but I've added some.

jelbourn commented on packages/core/schematics/migrations/renderer-to-renderer2/helpers.ts:
```diff
@@ -0,0 +1,392 @@
+function getAnyTypedParameter(
+    parameterName: string | ts.Identifier, isRequired = true): ts.ParameterDeclaration {
+  // Declare the parameter as `any` so we don't have to add extra logic
+  // to ensure that the generated code will pass type checking.
```
One trick we use inside Google during migrations is to introduce a new alias for `any` like `AnyDuringRendererMigration`, which more strongly encourages people to fix them later. This might actually be necessary when running this inside Google, since use of `any` is barred via lint rule.

devversion replied:
I agree that we need to encourage people to avoid the usage of `any` in general. The idea of having an alias for `AnyDuringRendererMigration` sounds like an idea. Ideally the parameters would be typed but I assume the problem was that it turns out overkill with managing imports right? 

<sup>FWIW: I have something implemented in other migrations that allows easy import management w/ re-using of existing imports, namespace imports etc.</sup>

jelbourn commented on packages/core/schematics/migrations/renderer-to-renderer2/helpers.ts:
```diff
@@ -0,0 +1,392 @@
+
+/** Names of the helper functions that are supported for this migration. */
+export const enum HelperFunction {
+  createElement = '__rendererCreateElementHelper',
```
Would it make sense to prefix these with `ng` instead of `__` to make it clear where they came from? I feel like using `__` is blurring the lines w/ TypeScript's helpers a bit too much

jelbourn commented on packages/core/schematics/migrations/renderer-to-renderer2/google3/rendererToRenderer2Rule.ts:
```diff
@@ -0,0 +1,121 @@
+    if (!rendererImport) {
+      return failures;
+    }
+
```
It might be good to add a comment here that explains the high-level strategy for how this schematic works

jelbourn commented on packages/core/schematics/migrations/renderer-to-renderer2/index.ts:
```diff
@@ -0,0 +1,120 @@
+      return;
+    }
+
+    const {typedNodes, methodCalls} = findRendererReferences(sourceFile, typeChecker);
```
It looks like there's some duplication between here and `rendererToRenderer2Rule.ts`. Could more of this be made common?

devversion replied:
This is a common issue with our migrations right now. I have some new approach for creating these TSLint/CLI migration drivers that avoid as much duplication as possible, but I'm not sure if it is in scope for this migration. See: #30956

devversion commented on packages/core/schematics/migrations/renderer-to-renderer2/util.ts:
```diff
@@ -0,0 +1,73 @@
+/** Checks whether a node is referring to the `Renderer`. */
+function isRendererReference(typeChecker: ts.TypeChecker, node: ts.Node): boolean {
+  const symbol = typeChecker.getTypeAtLocation(node).getSymbol();
+  return !!symbol && symbol.getName() === 'Renderer';
```
Could this possibly be `true` for other types named `Renderer` which aren't necessarily from `@angular/core`. There are two options to verify the origin:

1) Get the local symbol declaration and check if it comes from the `ImportSpecifier` in the angular core import.
2) Determine the source file location of the node and compare it

crisbeto replied:
I don't think we'll hit this for files that don't have an import to the `Renderer`.

devversion replied:
but isn't that what I'm saying? Based on my understanding this check will also resolve to true when someone imports a different `Renderer` that is not from `@angular/core`.

crisbeto replied:
Changed.

devversion commented on packages/core/schematics/migrations/renderer-to-renderer2/index.ts:
```diff
@@ -0,0 +1,120 @@
+    update.insertRight(
+        rendererImport.getStart(),
+        printer.printNode(
+            ts.EmitHint.Unspecified, replaceImport(rendererImport, 'Renderer', 'Renderer2'),
```
I think the `findRendererImport` util should return the actual import specifier. That way we don't need to re-print the whole named bindings (and throw off the formatting)

crisbeto replied:
I did it this way to handle the case where somebody might have both `Renderer` and `Renderer2` in the same `NamedImports`.

devversion replied:
That's a fair point, but can't we determine this before and only do the replacement if we need to update the import specifier. I'm not sure if there is a need to replace the whole named binding (as it breaks previous formatting when using the printer)

crisbeto replied:
I'm not sure that it's worth all the extra complexity to handle both cases. The way the TS printer prints it out by default (with spaces inside the braces) follows Angular's style guide anyway, and for the cases where the user's formatting is different, it's trivial to correct manually.

devversion commented on packages/core/schematics/migrations/renderer-to-renderer2/index.ts:
```diff
@@ -0,0 +1,120 @@
+      } else if (call.parent && ts.isExpressionStatement(call.parent)) {
+        // Otherwise if the call is inside an expression statement, drop the
+        // entire statement. This takes care of any trailing semicolons.
+        update.remove(call.parent.getStart(), call.parent.getWidth());
```
Should we add a todo / print something out if we just drop a node as part of the migration?

devversion replied:
Fair enough. :+1: maybe worth adding as a small comment here?

crisbeto replied:
Expanded the comment to mention the reasoning.

devversion commented on packages/core/schematics/migrations/renderer-to-renderer2/migration.ts:
```diff
@@ -0,0 +1,265 @@
+                                                args[2].getText() === 'undefined'))) {
+    // If we've got a call with two arguments, or one with three arguments where the last one is
+    // `undefined` or `null`, we can safely switch to a `removeStyle` call.
+    const innerExpression = (node.expression as ts.PropertyAccessExpression).expression;
```
The call expression is guaranteed to be a property access expression (as this is a search criteria when walking through the source file). Should there be a combined type that avoids this cast? Something similar to:

https://github.com/angular/angular/blob/c596795e6475524e53418653cdabeeed2d25852d/packages/core/schematics/utils/ng_decorators.ts#L12-L14

devversion replied:
I think you only need to cast it once when passing it to `migrateExpression`. Then all other migration utility functions are type-safe and the confusing cast (where it looks like it's unguarded) can be avoided.

crisbeto commented:
@kara @jelbourn @devversion I've addressed all of the feedback. Can you take another look?

alexeagle commented:
I'll tsunami this over google3 code now to look for broken corner cases

alexeagle commented:
This only affected 36 files at Google, so it feels like we don't need to polish the migration that much in order to land it in g3. A bunch of these could just be fixed manually in the couple places they occur, but I'll list them anyway:

1. in google3 we have a global type `AnyDuringAngularIvyMigration` - we shouldn't create types inline like 
![image](https://user-images.githubusercontent.com/47395/59800705-2a13b980-929c-11e9-8f5a-a35a4990a2b6.png)
(also this added type isn't referenced in the file)

2. Here's an error that the forwardRef return type isn't updated:
![image](https://user-images.githubusercontent.com/47395/59800897-98f11280-929c-11e9-9dbb-66c21f7b8861.png)

3. this is a nit: we added conditional logic here for `size` even though the typechecker could have told us that it's non-null in this location
![image](https://user-images.githubusercontent.com/47395/59802599-297c2280-929e-11e9-9226-b14d49c3f8ef.png)
and here
![image](https://user-images.githubusercontent.com/47395/59802661-4ca6d200-929e-11e9-80fd-a392fc8e4f9a.png)
and here
![image](https://user-images.githubusercontent.com/47395/59802717-66481980-929e-11e9-9640-441057622d69.png)
![image](https://user-images.githubusercontent.com/47395/59802807-9ee7f300-929e-11e9-9e9a-7697467abdc1.png)
![image](https://user-images.githubusercontent.com/47395/59802871-c048df00-929e-11e9-9298-7adbf80c4326.png)


4. A comment got duplicated:
![image](https://user-images.githubusercontent.com/47395/59802764-84157e80-929e-11e9-98c2-93d855f821f2.png)

alexeagle commented:
one more: a reference to `ngRendererCreateElementHelper` was added but the symbol isn't imported

![image](https://user-images.githubusercontent.com/47395/59804857-65b28180-92a4-11e9-8d91-011243fc34cd.png)

alexeagle commented:
I'll do a global presubmit run across google3 where I hide the `Renderer` symbol to verify the schematic caught all usages.

One thing we need to do in order to land this is update `packages/platform-browser-dynamic/src/compiler_reflector.ts` which currently has a reference to `Renderer` - can we remove that in this PR as well? Maybe it has broader implications and should be done separately; I'm not sure

kara commented:
@alexeagle Let's land the platform-browser change in a follow-up?

kara commented on packages/core/schematics/migrations/renderer-to-renderer2/index.ts:
```diff
@@ -0,0 +1,126 @@
+    const logger = context.logger;
+
+    logger.info('------ Renderer to Renderer2 Migration ------');
+    logger.info('As of Angular 9 the Renderer class is no longer available.');
```
```suggestion
    logger.info('As of Angular 9, the Renderer class is no longer available.');
```

kara commented on packages/core/schematics/migrations/renderer-to-renderer2/index.ts:
```diff
@@ -0,0 +1,126 @@
+
+    logger.info('------ Renderer to Renderer2 Migration ------');
+    logger.info('As of Angular 9 the Renderer class is no longer available.');
+    logger.info('Renderer2 should be used instead');
```
```suggestion
    logger.info('Renderer2 should be used instead.');
```

alexeagle commented:
https://github.com/angular/angular/pull/31162 has the removal of Renderer from platform-browser-dynamic

alexeagle commented:
I ran a global presubmit across all of Google code and confirmed it's green: this migration found all the locations where Renderer was used.

crisbeto commented:
@kara can you take another look? I've updated it to incorporate yours and Paul's feedback. I've also added handling for `forwardRef` based on the result from the G3 presubmit, but I wasn't able to reproduce the case that Alex saw where the helpers weren't inserted in one file. 

Regarding the case where the type could be inferred so we don't generate a ternary: I wasn't able to figure out how to get the final type through the type checker for something like `value + 'px'`, but even in these cases we generate valid code which is slightly longer. These cases are trivial to clean up manually afterwards if they don't match the consumer's code style.

crisbeto committed:
fixup! feat(core): add automatic migration from Renderer to Renderer2

kara commented on packages/core/schematics/migrations/renderer-to-renderer2/helpers.ts:
```diff
@@ -0,0 +1,403 @@
+/** Names of the helper functions that are supported for this migration. */
+export const enum HelperFunction {
+  any = 'AnyDuringRendererMigration',
+  createElement = 'ngRendererCreateElementHelper',
```
Looks like this comment was missed last round?

https://github.com/angular/angular/pull/30936#discussion_r295570953

kara commented on packages/core/schematics/migrations/renderer-to-renderer2/helpers.ts:
```diff
@@ -0,0 +1,403 @@
+  throw new Error(`Unsupported helper called "${name}".`);
+}
+
+/** Creates a helper our custom `any` type during the migration. */
```
```suggestion
/** Creates a helper for custom `any` type during the migration. */
```

crisbeto committed:
fixup! feat(core): add automatic migration from Renderer to Renderer2

crisbeto commented:
Addressed the latest set of comments.

kara reviewed (approved):
LGTM

kara commented:
[presubmit](http://test/OCL:256216619:BASE:256216636:1562095814099:3c95117)

kara commented:
Merge-assistance: codefresh lagging
